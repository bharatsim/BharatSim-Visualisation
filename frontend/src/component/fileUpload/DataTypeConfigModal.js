import React, { useState } from 'react';
import { Box, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import Modal from '../../uiComponent/Modal';
import { createSchema } from '../../utils/fileUploadUtils';
import Dropdown from '../../uiComponent/Dropdown';
import { convertStringArrayToOptions } from '../../utils/helper';
import dataTypesMapping from '../../constants/dataTypesMapping';

const dropdownStyles = makeStyles(() => ({
  configDropdown: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
  },
}));

const styles = makeStyles((theme) => ({
  configContainer: {
    width: theme.spacing(140),
    height: theme.spacing(100),
    padding: theme.spacing(4),
  },
}));

function renderDropdown(columnName, value, onChange) {
  const classes = dropdownStyles();
  return (
    <Box className={classes.configDropdown} key={columnName}>
      <Typography>{columnName}</Typography>
      <Dropdown
        label="Select datatype"
        value={value}
        options={convertStringArrayToOptions(Object.values(dataTypesMapping))}
        id={columnName}
        onChange={(selectedValue) => onChange(columnName, selectedValue)}
      />
    </Box>
  );
}

function DataTypeConfigModal({ dataRow, isOpen, closeModal, onApply, onCancel }) {
  const schema = createSchema(dataRow);
  const [values, setValues] = useState(schema);
  const classes = styles();

  function onChange(key, value) {
    setValues((prevState) => ({ ...prevState, [key]: value }));
  }

  return (
    <Modal
      handleClose={closeModal}
      open={isOpen}
      title="Configure Datatype"
      actions={[
        { name: 'cancel', handleClick: onCancel, variant: 'outlined' },
        {
          name: 'apply and upload',
          handleClick: () => onApply(values),
          variant: 'contained',
          color: 'primary',
        },
      ]}
    >
      <Box className={classes.configContainer}>
        {Object.keys(schema).map((columnName) =>
          renderDropdown(columnName, values[columnName], onChange),
        )}
      </Box>
    </Modal>
  );
}

DataTypeConfigModal.propTypes = {
  dataRow: PropTypes.shape({}).isRequired,
  isOpen: PropTypes.bool.isRequired,
  closeModal: PropTypes.func.isRequired,
  onApply: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

export default DataTypeConfigModal;
