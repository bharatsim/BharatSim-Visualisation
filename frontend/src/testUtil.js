// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, within } from '@testing-library/react';

export function selectDropDownOption(container, dropDownId, optionId) {
  const dropDown = container.getByTestId(dropDownId);
  fireEvent.mouseDown(within(dropDown).getByRole('button'));
  const options = within(
    within(document.getElementById(`menu-${dropDownId}`)).getByRole('listbox'),
  );
  fireEvent.click(options.getByTestId(`${dropDownId}-${optionId}`));
}
