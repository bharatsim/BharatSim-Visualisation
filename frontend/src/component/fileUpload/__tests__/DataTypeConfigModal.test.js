import { render, within } from '@testing-library/react';
import React from 'react';
import { fireEvent } from '@testing-library/dom';
import DataTypeConfigModal from '../DataTypeConfigModal';
import { selectDropDownOption } from '../../../testUtil';

describe('DataTypeConfigModal', () => {
  const props = {
    dataRow: { col1: 1, col2: 'a', col3: '3' },
    isOpen: true,
    closeModal: jest.fn(),
    onApply: jest.fn(),
    onCancel: jest.fn(),
  };

  beforeEach(() => {
    render(<DataTypeConfigModal {...props} />);
  });

  it('should match a snapshot', () => {
    const container = document.querySelector('.MuiPaper-root');

    expect(container).toMatchSnapshot();
  });
  it('should close the modal on click of cancel', () => {
    const container = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = container;
    const cancelButton = getByText('cancel');

    fireEvent.click(cancelButton);

    expect(props.onCancel).toHaveBeenCalled();
  });
  it('should call apply function on click of apply and upload button with default schema', () => {
    const container = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = container;
    const applyButton = getByText('apply and upload');

    fireEvent.click(applyButton);

    expect(props.onApply).toHaveBeenCalledWith({ col1: 'Number', col2: 'String', col3: 'String' });
  });

  it('should call apply function with updated schema', () => {
    const container = within(document.querySelector('.MuiPaper-root'));
    const { getByText } = container;
    const applyButton = getByText('apply and upload');

    selectDropDownOption(container, 'col1', 'String');
    fireEvent.click(applyButton);

    expect(props.onApply).toHaveBeenCalledWith({ col1: 'String', col2: 'String', col3: 'String' });
  });
});
