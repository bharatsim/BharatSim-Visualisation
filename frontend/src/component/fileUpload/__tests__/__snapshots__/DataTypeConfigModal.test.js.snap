// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTypeConfigModal should match a snapshot 1`] = `
<div
  aria-labelledby="customized-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root Component-root-3"
    id="customized-dialog-title"
  >
    <div
      class="MuiBox-root MuiBox-root-6 Component-titleContainer-5"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Configure Datatype
      </h5>
      <button
        aria-label="close"
        class="MuiButtonBase-root MuiIconButton-root Component-closeButton-4"
        data-testid="button-icon-close"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiDialogContent-root WithStyles(ForwardRef(DialogContent))-root-7 MuiDialogContent-dividers"
  >
    <div
      class="MuiBox-root MuiBox-root-8 makeStyles-configContainer-1"
    >
      <div
        class="MuiBox-root MuiBox-root-9 makeStyles-configDropdown-2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          col1
        </p>
        <div
          class="MuiFormControl-root makeStyles-formControl-10"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
            data-shrink="true"
            id="dropdown-label"
          >
            Select datatype
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            data-testid="col1"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="dropdown-label col1"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
              id="col1"
              role="button"
              tabindex="0"
            >
              Number
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="Number"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-12 makeStyles-configDropdown-2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          col2
        </p>
        <div
          class="MuiFormControl-root makeStyles-formControl-10"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
            data-shrink="true"
            id="dropdown-label"
          >
            Select datatype
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            data-testid="col2"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="dropdown-label col2"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
              id="col2"
              role="button"
              tabindex="0"
            >
              String
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="String"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-13 makeStyles-configDropdown-2"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          col3
        </p>
        <div
          class="MuiFormControl-root makeStyles-formControl-10"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
            data-shrink="true"
            id="dropdown-label"
          >
            Select datatype
          </label>
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
            data-testid="col3"
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="dropdown-label col3"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
              id="col3"
              role="button"
              tabindex="0"
            >
              String
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              tabindex="-1"
              value="String"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiDialogActions-root WithStyles(ForwardRef(DialogActions))-root-14 MuiDialogActions-spacing"
  >
    <div
      class="MuiBox-root MuiBox-root-16 makeStyles-buttonGroup-15"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined Mui-focusVisible Mui-focusVisible"
        data-testid="button-cancel"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          cancel
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
            />
          </span>
        </span>
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-focusVisible Mui-focusVisible"
        data-testid="button-apply and upload"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          apply and upload
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
            />
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
